### PR TITLE
server/market: refactor trading limits

### PR DESCRIPTION
**Pull request by buck54321**
Add a per-market limit scaling factor, the parcel size. A "parcel" is the product of parcel size and lot size. Base user trading limits on parcels, not lots, and evaluate globally, not per-market.

User trading limits still scale with score and tier. The old InitTakerLotLimit, AbsTakerLotLimit, and BookedLotLimit system is replaced.

The docs from **dex/market.go** explains parcel limits a little better.
```Go
// Parcels are used to track user trading limits. A parcel is a number of
// lots. How many lots are in a parcel is defined as part of a market's
// configuration. Markets with low-fee assets might have very small lot
// sizes, so would set a large parcel size. A market with high-fee assets
// would have a larger lot size, so would set a small parcel size.

// Parcels are tracked accross all known markets, and trading limits are
// assessed globally.

// When calculating trading limits, each market will report its parcels
// for the user. There are no partial parcels for the purposes of
// calculation. If the parcel size is 10 lots, and the user has 11 lots
// booked. the market will report 2 parcels. This fact effectively limits
// the number of markets singly-tiered, score 0 users can trade on to
// PerTierBaseParcelLimit.

// Parcel limits are measured in units of "parcel weight", which is kind of
// like "effective quantity". For makers, parcel weight = quantity. Orders
// which are likely to be takers have their parcel weight doubled. The
// quantity in settling matches has the same weight as maker quantity.

// DefaultParcelSize is the default number of lots per parcel, configurable
// per market.
DefaultParcelSize = 4 // lots per parcel
// PerTierBaseParcelLimit is the number of parcels users can trade per tier.
PerTierBaseParcelLimit = 2
// ParcelLimitScoreMultiplier defines the maximum parcel limit multiplier
// that a user achieves with max score. The users parcel limit ranges from
// the starting value of tier*PerTierBaseParcelLimit to the maximum value of
// tier*PerTierBaseParcelLimit*ParcelLimitScoreMultiplier. By default,
// singly-tiered users have max lot limits limited to the range (8,24).
ParcelLimitScoreMultiplier = 3
```

### CODE REVIEW

**Review by JoeGruffins — COMMENTED** [review]
These changes seem to have stopped the loadbots with trading limits. Running like always will not be able to trade:
```
./loadbot -p sidestacker -mkt btc_eth -oscillate -trace -randomosc
```

Would like to keep them working, maybe start with higher bond tiers?

File: dex/market.go [hunk 1]
```diff
@@ -13,6 +13,39 @@ const (
...
+
+	// Parcels are tracked accross all known markets, and trading limits are
+	// assessed globally.
+
+	// When calculating trading limits, each market will report its parcels
+	// for the user. There are no partial parcels for the purposes of
+	// calculation. If the parcel size is 10 lots, and the user has 11 lots
+	// booked. the market will report 2 parcels. This fact effectively limits
+	// the number of markets singly-tiered, score 0 users can trade on to
+	// PerTierBaseParcelLimit.
```
Comment (buck54321) [inline dex/market.go:31]: Although I am open to using partial parcels too using floats.

**Comment by buck54321** [comment]
> These changes seem to have stopped the loadbots with trading limits. Running like always will not be able to trade:
> 
> ```
> ./loadbot -p sidestacker -mkt btc_eth -oscillate -trace -randomosc
> ```
> 
> Would like to keep them working, maybe start with higher bond tiers?

Either increase bond tier (my preference), or you could add a larger parcel limit to the markets in the simnet harness.

Loadbot should really just do the math needed to ensure it stays under it's limit or sets its tier high enough, but the client will provide some utility for these calculations soon too, so it might be prudent to wait on that.

**Comment by buck54321** [comment]
> Either increase bond tier (my preference), or you could add a larger parcel limit to the markets in the simnet harness.
> 
> Loadbot should really just do the math needed to ensure it stays under it's limit or sets its tier high enough, but the client will provide some utility for these calculations soon too, so it might be prudent to wait on that.

I don't know why I'm telling you to do it. I'll take care of it.

File: server/market/orderrouter.go [hunk 2]
```diff
@@ -661,7 +674,69 @@ func (r *OrderRouter) sufficientAccountBalance(accountAddr string, ord order.Ord
...
+		}
+	}
+	targetMarketParcels := calcParcels(settlingQty)
+
+	return otherMarketParcels+targetMarketParcels <= parcelLimit
+}
+
 func (r *OrderRouter) submitOrderToMarket(tunnel MarketTunnel, oRecord *orderRecord) *msgjson.Error {
+	oRecord.checkParcels = func(f marketParcelCalculator) bool {
+		return r.checkParcelLimit(oRecord, f)
```
[the remark this answers is not in the document]
Comment (buck54321) [reply server/market/orderrouter.go]: Done.

File: dex/market.go (hunk 1 above)
Comment (buck54321) [reply dex/market.go:31]: I changed it to floats.

**Comment by buck54321** [comment]
Loadbot updated to bond to tier 20, and bond amounts and parcel sizes adjusted in the simnet harness.

**Review by JoeGruffins — COMMENTED** [review]
Running loadbot I still see a lot of `2023-09-25 15:04:03.932 [INF] MANTLE:STACKER:0: Over-limit error. Order not placed.` Some orders do make it to market but not all.

**Comment by buck54321** [comment]
> Running loadbot I still see a lot of `2023-09-25 15:04:03.932 [INF] MANTLE:STACKER:0: Over-limit error. Order not placed.` Some orders do make it to market but not all.

Raised the tier to 100, which works out to about 50 DCR. That seems reasonable for a high-volume bot, though I question whether the sidestacker **really** needs to have so many active orders.

**Review by JoeGruffins — APPROVED** [review]

**Review by martonp — APPROVED** [review]

**Pull request merged**